spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/LocalizedQueueConnectionFactory.java

```diff
@@ -0,0 +1,358 @@
+			String[] addresses, String[] adminUris, String[] nodes, String vhost,
+			String username, String password, boolean useSSL, String keyStore, String trustStore,
+			String keyStorePassPhrase, String trustStorePassPhrase) {
+		Assert.notNull(defaultConnectionFactory, "'defaultConnectionFactory' cannot be null");
+		Assert.isTrue(addresses.length == nodes.length,
+				"'addresses', 'adminAddresses', and 'nodes' properties must have equal length");
+		this.defaultConnectionFactory = defaultConnectionFactory;
+		this.adminUris = Arrays.copyOf(adminUris, adminUris.length);
+		for (int i = 0; i < addresses.length; i++) {
+			this.nodeToAddress.put(nodes[i], addresses[i]);
+		}
+		this.vhost = vhost;
+		this.username = username;
+		this.password = password;
+		this.useSSL = useSSL;
+		this.sslPropertiesLocation = null;
+		this.keyStore = keyStore;
+		this.trustStore = trustStore;
+		this.keyStorePassPhrase = keyStorePassPhrase;
+		this.trustStorePassPhrase = trustStorePassPhrase;
+	}
+
+	@Override
+	public Connection createConnection() throws AmqpException {
+		return this.defaultConnectionFactory.createConnection();
+	}
+
+	@Override
+	public String getHost() {
+		return this.defaultConnectionFactory.getHost();
+	}
+
+	@Override
+	public int getPort() {
+		return this.defaultConnectionFactory.getPort();
+	}
+
+	@Override
+	public String getVirtualHost() {
+		return this.vhost;
+	}
+
+	@Override
+	public void addConnectionListener(ConnectionListener listener) {
+		this.defaultConnectionFactory.addConnectionListener(listener);
+	}
+
+	@Override
+	public boolean removeConnectionListener(ConnectionListener listener) {
+		return this.defaultConnectionFactory.removeConnectionListener(listener);
+	}
+
+	@Override
+	public void clearConnectionListeners() {
+		this.defaultConnectionFactory.clearConnectionListeners();
+	}
+
+	@Override
+	public ConnectionFactory getTargetConnectionFactory(Object key) {
+		String queue = ((String) key);
+		queue = queue.substring(1, queue.length() - 1);
+		Assert.isTrue(!queue.contains(","), "Cannot use LocalizedQueueConnectionFactory with more than one queue: " + key);
+		ConnectionFactory connectionFactory = determineConnectionFactory(queue);
+		if (connectionFactory == null) {
+			return this.defaultConnectionFactory;
+		}
+		else {
+			return connectionFactory;
+		}
+	}
+
+	private ConnectionFactory determineConnectionFactory(String queue) {
+		for (int i = 0; i < this.adminUris.length; i++) {
+			String adminUri = this.adminUris[i];
+			if (!adminUri.endsWith("/api/")) {
+				adminUri += "/api/";
+			}
+			try {
+				Client client = createClient(adminUri, this.username, this.password);
+				QueueInfo queueInfo = client.getQueue(this.vhost, queue);
+				if (queueInfo != null) {
+					String node = queueInfo.getNode();
+					if (node != null) {
+						String uri = this.nodeToAddress.get(node);
+						if (uri != null) {
+							return nodeConnectionFactory(queue, node, uri);
```

Comment:
Why don't we cache `queue -> connectionFactory` as well?
I just don't like that `client.getQueue()` HTTP request for each message...

Or is this fully intentional because of HA failover aim?